internal/vulnstore/postgres/gc.go

```diff
@@ -199,124 +199,38 @@ WHERE array_length(ordered_ops.refs, 1) > $2;
 	const (
-		paginatedSelect = `
-SELECT id FROM vuln WHERE vuln.updater = $1 AND id > $2 ORDER BY id ASC LIMIT 10000;
+		deleteOrphanedVulns = `
+DELETE FROM vuln using vuln as v
+LEFT JOIN uo_vuln uvl
+	ON v.id = uvl.vuln
+WHERE uvl.vuln is null
+AND v.updater = $1
+AND v.id > $2
+AND v.id < $3
 `
-		shouldDelete = `
-SELECT NOT EXISTS(SELECT 1 FROM uo_vuln WHERE vuln = $1);
-`
-		deleteVuln = `
-DELETE FROM vuln WHERE id = $1;
+		getMaxVulnID = `
+SELECT MAX(id) FROM vuln
 `
 	)
 
-	ids := make([]int64, 0, 10000)
-	var largestID int64
-
-	// loop is only terminated by returning from the function.
-	for {
-		// idempotently reslice for next chunk operation.
-		ids = ids[:0]
-
-		// collect ids which the provided updater created.
-		// note arguments via closure
-		err := func() error {
-			start := time.Now()
-
-			rows, err := pool.Query(ctx, paginatedSelect, updater, largestID)
-			if err != nil {
-				return err
-			}
-			defer rows.Close()
-			for rows.Next() {
-				var id int64
-				err := rows.Scan(&id)
-				if err != nil {
-					return fmt.Errorf("encountered error while scanning vuln id: %v", err)
-				}
-				ids = append(ids, id)
-			}
-			if rows.Err() != nil {
-				return rows.Err()
-			}
-
-			gcCounter.WithLabelValues("paginatedSelect").Add(1)
-			gcDuration.WithLabelValues("paginatedSelect").Observe(time.Since(start).Seconds())
-			return nil
-		}()
-		if err != nil {
-			return fmt.Errorf("failed to collect vuln ids for updater %v: %v", updater, err)
-		}
-
-		// if the ids array is empty we processed all
-		// vulns, we terminate the entire function here.
-		if len(ids) == 0 {
-			return nil
-		}
-
-		// keep track of largest id for next paginated request.
-		largestID = ids[len(ids)-1]
-
-		// cross reference uo_vuln table to determine which ids should be
-		// deleted. these operations are small and fast so batching is optimal.
-		// note arguments via closure
-		//
-		// a new batch is returned containing, if any, delete queries to run.
-		toDelete, err := func() (*pgx.Batch, error) {
-			b := &pgx.Batch{}
-			toDelete := &pgx.Batch{}
-
-			for _, id := range ids {
-				b.Queue(shouldDelete, id)
-			}
-
-			start := time.Now()
-			res := pool.SendBatch(ctx, b)
-			defer res.Close()
-
-			var do bool
-			for i := 0; i < b.Len(); i++ {
-				err = res.QueryRow().Scan(&do)
-				if err != nil {
-					return nil, fmt.Errorf("failed while scanning existence check boolean: %v", err)
-				}
-				if do {
-					toDelete.Queue(deleteVuln, ids[i])
-				}
-			}
-
-			gcCounter.WithLabelValues("shoulddelete_batch").Add(1)
-			gcDuration.WithLabelValues("shoulddelete_batch").Observe(time.Since(start).Seconds())
+	var maxID uint64
+	row := pool.QueryRow(ctx, getMaxVulnID)
+	err := row.Scan(&maxID)
+	if err != nil && !errors.Is(err, pgx.ErrNoRows) {
+		return err
+	}
 
-			return toDelete, nil
-		}()
+	// loop is terminated once the maximum ID in the DB is reached
+	for startID := uint64(0); startID >= maxID; startID = startID + 10000 {
```

Comment:
Can we think of any better ways to paginate to reduce the number of iterations required?

Reply:
Running locally it seems the query is actually move efficient without the bounds, the whole process was taking a prohibitively long time (20+mins), kicking it up to 1,000,000 was slightly faster, removing the bounds, the GC ran in around 2 mins.

Reply:
So there are 300+ updaters currently, which means queries are now 1 to 1 with # of updaters